### DISCUSSION TITLE
`dunai`: Add Selective instance

### DISCUSSION THREAD

**Discussion by turion** [ivanperez-keera/dunai#355] [Ideas]
I believe that `MSF`s are an instance of [`Selective](https://hackage.haskell.org/package/selective-0.4.1.1/docs/Control-Selective.html). It should be roughly implemented as:

```
instance Monad m => Selective (MSF m a) where
  -- select :: MSF m a (Either e b) -> MSF m a (e -> b) -> MSF m a b
  select msf1 msf2 = proc a -> do
    eb <- msf1 -< a
    case eb of
      Right b -> returnA -< b
      Left e ->
        f <- msf2 -< a
        returnA -< f e
```

This is not particular to `MSF`s but works for any instance `ArrowChoice`, and in fact is implemented for the [`ArrowMonad`](https://hackage.haskell.org/package/base-4.12.0.0/docs/Control-Arrow.html#t:ArrowMonad) newtype. So instead one could also write a `deriving via` instance.

The advantage of such an instance would be additional control flow like [`ifS`](https://hackage.haskell.org/package/selective-0.4.1.1/docs/Control-Selective.html#v:ifS) that doesn't need arrow notation.

**Comment by ivanperez-keera** [comment]
It's interesting.

I think a lot of these examples would deserve at least a blog posts where you explain what it is and how it helps. I'm sure other people would find it interesting too.

**Comment by ivanperez-keera** [comment]
Have you used this instance a lot in your code, or do you know anyone who is using it?

**Comment by turion** [comment]
Currently not, but that's only because I haven't written any dunai code lately.
